File: checkpoints/go/02-streaming/main.go
package main

import (
	"context"
	"fmt"

	copilot "github.com/github/copilot-sdk/go"
)

func streamResponse(session *copilot.Session, prompt string) error {
	unsubscribe := session.On(func(event copilot.SessionEvent) {
		if delta, ok := event.Data.(*copilot.AssistantMessageDeltaData); ok {
			fmt.Print(delta.DeltaContent)
		}
	})
	defer unsubscribe()

	_, err := session.SendAndWait(context.Background(), copilot.MessageOptions{Prompt: prompt})
	fmt.Println()
	return err
}

func main() {
	client := copilot.NewClient(&copilot.ClientOptions{LogLevel: "error"})
	if err := client.Start(context.Background()); err != nil {
		panic(err)
	}
	defer client.Stop()

	session, err := client.CreateSession(context.Background(), &copilot.SessionConfig{
		Streaming: copilot.Bool(true),
	})
	if err != nil {
		panic(err)
	}
	defer session.Disconnect()

	if err := streamResponse(session, "Explain accessible names in three short bullet points."); err != nil {
		panic(err)
	}
}

File: checkpoints/rust/02-streaming/src/main.rs
use std::io::{self, Write};

use github_copilot_sdk::types::{MessageOptions, SessionConfig};
use github_copilot_sdk::{Client, ClientOptions};

#[tokio::main]
async fn main() -> Result<(), github_copilot_sdk::Error> {
    let client = Client::start(ClientOptions::default()).await?;
    let mut config = SessionConfig::default();
    config.streaming = Some(true);
    let session = client.create_session(config).await?;

    let mut events = session.subscribe();
    tokio::spawn(async move {
        while let Ok(event) = events.recv().await {
            if event.event_type == "assistant.message_delta" {
                if let Some(delta) = event.data.get("deltaContent").and_then(|value| value.as_str()) {
                    print!("{delta}");
                    io::stdout().flush().ok();
                }
            }
        }
    });

    session
        .send_and_wait(MessageOptions::new(
            "Explain accessible names in three short bullet points.",
        ))
        .await?;
    println!();
    session.disconnect().await?;
    client.stop().await?;
    Ok(())
}
